File: packages/expo-liquid-glass/ios/ExpoLiquidGlassModule.swift
import ExpoModulesCore

/**
 Native side of `<LiquidGlassView>` on iOS.

 On iOS 26+ this hosts Apple's real `UIGlassEffect` inside a
 `UIVisualEffectView`; the system performs its own backdrop sampling,
 refraction and highlighting, so the Android optical tuning props are accepted
 as no-ops. Below iOS 26 it falls back to an ultra-thin material blur.
 */
public class ExpoLiquidGlassModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoLiquidGlass")

    View(LiquidGlassView.self) {
      Prop("cornerRadius") { (view: LiquidGlassView, radius: Double?) in
        view.setCornerRadius(radius)
      }
      Prop("tint") { (view: LiquidGlassView, color: UIColor?) in
        view.setTint(color)
      }
      Prop("interactive") { (view: LiquidGlassView, interactive: Bool?) in
        view.setInteractive(interactive ?? false)
      }

      // Android-only tuning props, accepted so shared JS just works.
      Prop("providerId") { (_: LiquidGlassView, _: String?) in }
      Prop("blurRadius") { (_: LiquidGlassView, _: Double?) in }
      Prop("refractionHeight") { (_: LiquidGlassView, _: Double?) in }
      Prop("refractionAmount") { (_: LiquidGlassView, _: Double?) in }
      Prop("saturation") { (_: LiquidGlassView, _: Double?) in }
      Prop("chromaticAberration") { (_: LiquidGlassView, _: Double?) in }
      Prop("noiseAlpha") { (_: LiquidGlassView, _: Double?) in }
      Prop("highlightAlpha") { (_: LiquidGlassView, _: Double?) in }
      Prop("highlightWidth") { (_: LiquidGlassView, _: Double?) in }
      Prop("lightAngle") { (_: LiquidGlassView, _: Double?) in }
    }
  }
}

class LiquidGlassView: ExpoView {
  private let effectView = UIVisualEffectView(effect: nil)
  private var isInteractiveGlass = false
  private var glassTint: UIColor?
  /// nil = capsule (radius follows bounds); a number = fixed radius in points.
  private var explicitCornerRadius: Double?

  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    clipsToBounds = true
    effectView.frame = bounds
    effectView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
    addSubview(effectView)
    applyEffect()
  }

  override func didAddSubview(_ subview: UIView) {
    super.didAddSubview(subview)
    // React Native children always render above the glass.
    if subview !== effectView {
      sendSubviewToBack(effectView)
    }
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    let radius: CGFloat
    if let explicit = explicitCornerRadius {
      radius = min(CGFloat(explicit), min(bounds.width, bounds.height) / 2)
    } else {
      radius = min(bounds.width, bounds.height) / 2
    }
    layer.cornerRadius = radius
    layer.cornerCurve = .continuous
  }

  func setCornerRadius(_ radius: Double?) {
    explicitCornerRadius = radius
    setNeedsLayout()
  }

  func setTint(_ color: UIColor?) {
    glassTint = color
    applyEffect()
  }

  func setInteractive(_ interactive: Bool) {
    isInteractiveGlass = interactive
    applyEffect()
  }

  private func applyEffect() {
    if #available(iOS 26.0, *) {
      let glass = UIGlassEffect()
      glass.isInteractive = isInteractiveGlass
      if let tint = glassTint {
        glass.tintColor = tint
      }
      effectView.effect = glass
    } else {
      effectView.effect = UIBlurEffect(style: .systemUltraThinMaterial)
      effectView.backgroundColor = glassTint?.withAlphaComponent(
        (glassTint?.cgColor.alpha ?? 0) * 0.4
      )
    }
  }
}

File: packages/liquid-glass-kit/android/src/main/java/expo/modules/liquidglass/ExpoLiquidGlassModule.kt
package expo.modules.liquidglass

import android.content.Context
import android.graphics.Canvas
import android.view.MotionEvent
import dev.liquidglass.view.CORNER_RADIUS_CAPSULE
import dev.liquidglass.view.GlassViewController
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.views.ExpoView

/**
 * Native side of `<LiquidGlassView>`: the glass engine drawn directly on the
 * React Native view (children render above it). The view embeds
 * [GlassViewController] rather than wrapping a child view so React Native's
 * mounting (Paper and Fabric) keeps full ownership of the child index space.
 */
class ExpoLiquidGlassModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("ExpoLiquidGlass")

        View(LiquidGlassExpoView::class) {
            Prop("providerId") { view: LiquidGlassExpoView, id: String ->
                view.providerId = id
            }
            Prop("cornerRadius") { view: LiquidGlassExpoView, radius: Float? ->
                view.glass.cornerRadiusDp = radius ?: CORNER_RADIUS_CAPSULE
            }
            Prop("blurRadius") { view: LiquidGlassExpoView, radius: Float? ->
                view.glass.blurRadiusDp = radius ?: 20f
            }
            Prop("refractionHeight") { view: LiquidGlassExpoView, height: Float? ->
                view.glass.refractionHeightDp = height ?: 10f
            }
            Prop("refractionAmount") { view: LiquidGlassExpoView, amount: Float? ->
                view.glass.refractionAmountDp = amount ?: 12f
            }
            Prop("saturation") { view: LiquidGlassExpoView, saturation: Float? ->
                view.glass.saturation = saturation ?: 1.5f
            }
            Prop("tint") { view: LiquidGlassExpoView, color: Int? ->
                view.glass.glassTintColor = color ?: 0
            }
            Prop("chromaticAberration") { view: LiquidGlassExpoView, amount: Float? ->
                view.glass.chromaticAberration = (amount ?: 0f).coerceIn(0f, 1f)
            }
            Prop("noiseAlpha") { view: LiquidGlassExpoView, alpha: Float? ->
                view.glass.noiseAlpha = (alpha ?: 0.015f).coerceIn(0f, 1f)
            }
            Prop("highlightAlpha") { view: LiquidGlassExpoView, alpha: Float? ->
                view.glass.highlightAlpha = (alpha ?: 0.45f).coerceIn(0f, 1f)
            }
            Prop("highlightWidth") { view: LiquidGlassExpoView, width: Float? ->
                view.glass.highlightWidthDp = width ?: 2f
            }
            Prop("lightAngle") { view: LiquidGlassExpoView, degrees: Float? ->
                view.glass.lightAngleDegrees = degrees ?: 245f
            }
            Prop("interactive") { view: LiquidGlassExpoView, interactive: Boolean? ->
                view.glass.isGlassInteractive = interactive ?: false
            }
        }
    }
}

class LiquidGlassExpoView(
    context: Context,
    appContext: AppContext,
) : ExpoView(context, appContext) {

    /** The embedded glass engine; props from JS are applied to it directly. */
    val glass = GlassViewController(this)

    var providerId: String = "default"
        set(value) {
            if (field == value && glass.provider != null) return
            field = value
            if (isAttachedToWindow) LiquidGlassRegistry.resolve(value, glass)
        }

    init {
        setWillNotDraw(false)
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        LiquidGlassRegistry.resolve(providerId, glass)
        glass.onHostAttached()
    }

    override fun onDetachedFromWindow() {
        LiquidGlassRegistry.cancelWaiting(glass)
        glass.onHostDetached()
        super.onDetachedFromWindow()
    }

    override fun dispatchTouchEvent(event: MotionEvent): Boolean {
        glass.handleTouch(event)
        return super.dispatchTouchEvent(event)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        glass.draw(canvas)
    }
}
